attacks-and-mitigations.md

```diff
@@ -591,45 +606,50 @@ mechanisms described next.
-check fails: the client uses its correct verifier, but the code
-is associated with a challenge that does not match this verifier. PKCE is a deployed
-OAuth feature, although its originally intended use was solely focused
-on securing native apps, not the broader use recommended by this
-document.
-    
+The PKCE mechanism specified in [@!RFC7636] can be used as a countermeasure.
+When the attacker attempts to inject an authorization code, the check of the
+`code_verifier` fails: the client uses its correct verifier, but the code is
+associated with a `code_challenge` that does not match this verifier. PKCE is a
+deployed OAuth feature, although its originally intended use was solely focused
+on securing native apps, not the broader use recommended by this document.
+
+PKCE does not only protect against the autorization code injection attack, but
+also protects authorization codes created for public clients: PKCE ensures that
+an attacker cannot redeem a stolen authorization code at the token endpoint of
+the authorization server without knowledge of the `code_verifier`. 
     
 #### Nonce {#nonce_as_injection_protection}
 
-OpenID Connect's existing `nonce` parameter can be used for the same
-purpose. The `nonce` value is one-time use and created by the client.
-The client is supposed to bind it to the user agent session and sends
-it with the initial request to the OpenID Provider (OP). The OP binds
-`nonce` to the authorization code and attests this binding in the ID
-Token, which is issued as part of the code exchange at the token
-endpoint. If an attacker injected an authorization code in the
-authorization response, the nonce value in the client session and the
-nonce value in the ID token will not match and the attack is detected.
-The assumption is that an attacker cannot get hold of the user agent
-state on the victim's device, where he has stolen the respective
-authorization code.
-
-It is important to note that this countermeasure only works if the
-client properly checks the `nonce` parameter in the ID Token and does
-not use any issued token until this check has succeeded. More
-precisely, a client protecting itself against code injection using the
-`nonce` parameter,
-
-  1. MUST validate the `nonce` in the ID Token obtained from the
-     token endpoint, even if another ID Token was obtained from the
-     authorization response (e.g., `response_type=code+id_token`), and
-  1. MUST ensure that, unless and until that check succeeds, all
-     tokens (ID Tokens and the access token) are disregarded and not
-     used for any other purpose.
+OpenID Connect's existing `nonce` parameter can protect against authorization
+code injection attacks. The `nonce` value is one-time use and created by the
+client. The client is supposed to bind it to the user agent session and sends it
+with the initial request to the OpenID Provider (OP). The OP binds `nonce` to
+the authorization code and attests this binding in the ID Token, which is issued
```

Comment:
The way I think about it, the OP doesn't 'bind' anything with the nonce. It just puts the nonce value it received into the ID token. It's just a bit of state that is carried though. The client does the binding and checking.

Reply:
I agree. The wording is slightly strange. Will fix.